### ISSUE TITLE
OpenAPI endpoint gives validation error

### ISSUE THREAD

**Issue by justb4** [geopython/pygeoapi#168]
Came out of investigation for a [WFS3 GeoHealthCheck plugin](https://github.com/geopython/GeoHealthCheck/issues/221).

Interesting, there is quite some OpenAPI Python stuff out there, like a Spec Validator:
https://github.com/p1c2u/openapi-spec-validator

Very easy to run, but gives validation errors on the `pygeoapi` `master` demo WFS3 I try:

```
$ pip install openapi-spec-validator
$ python
>>> from openapi_spec_validator import validate_spec_url
>>> validate_spec_url('https://demo.pygeoapi.io/master/api?f=json')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/just/.pyenv/versions/geohealthcheck2.7/lib/python2.7/site-packages/openapi_spec_validator/shortcuts.py", line 16, in validate
    return validator_callable(spec, spec_url=url)
  File "/Users/just/.pyenv/versions/geohealthcheck2.7/lib/python2.7/site-packages/openapi_spec_validator/validators.py", line 48, in validate
    raise err
openapi_spec_validator.exceptions.OpenAPIValidationError: {'content': {'application/json': {'example': {'inputs': [{'type': 'text/plain', 'id': 'name', 'value': 'hi there'}]}, 'schema': {'x-scope': ['https://demo.pygeoapi.io/master/api?f=json'], '$ref': 'https://raw.githubusercontent.com/opengeospatial/wps-rest-binding/master/core/openapi/schemas/execute.yaml'}}}, 'required': True, 'description': 'Mandatory execute request JSON'} is not valid under any of the given schemas

Failed validating 'oneOf' in schema['properties']['paths']['patternProperties']['^/']['properties']['post']['properties']['requestBody']:
    {'oneOf': [{'$ref': '#/definitions/requestBody'},
               {'$ref': '#/definitions/reference'}]}

On instance['paths']['/processes/hello-world/jobs']['post']['requestBody']:
    {'content': {'application/json': {'example': {'inputs': [{'id': 'name',
                                                              'type': 'text/plain',
                                                              'value': 'hi there'}]},
                                      'schema': {'$ref': 'https://raw.githubusercontent.com/opengeospatial/wps-rest-binding/master/core/openapi/schemas/execute.yaml',
                                                 'x-scope': ['https://demo.pygeoapi.io/master/api?f=json']}}},
     'description': 'Mandatory execute request JSON',
     'required': True}

```

Has to do with `Processes`. If I leave out `/processes/hello-world/jobs`, then validates ok:

```
  /processes/hello-world/jobs:
    get:
      description: Hello World process
.
.
      responses:
        200:
          description: successful operation
        400:
          description: Invalid id supplied
        404:
          description: not found
      summary: Process Hello World process execution
      tags:
      - hello-world
```
Some other, not all, WFS3 endpoints validate ok like:

```
>>> validate_spec_url('https://www.ldproxy.nrw.de/kataster/api/?f=json')
>>> validate_spec_url('http://beta.fmi.fi/data/3/wfs/sofp/api.json')
```

**Comment by justb4** [comment]
Ok nailed it down to this line:

```
            schema:
              $ref: https://raw.githubusercontent.com/opengeospatial/wps-rest-binding/master/core/openapi/schemas/execute.yaml
```

Here my OpenAPI knowledge lacks... Anyone from @pvgenuchten @tomkralidis ?

**Comment by tomkralidis** [comment]
Not sure.  The OpenAPI document works in https://editor.swagger.io fine

**Comment by justb4** [comment]
Maybe the validator https://github.com/p1c2u/openapi-spec-validator is more strict?

At least it does not like the [execute.yaml](https://raw.githubusercontent.com/opengeospatial/wps-rest-binding/master/core/openapi/schemas/execute.yaml) schema. Also to me it looks a bit "strange" (though may be lack of understanding):

* the `hello-world` Process metadata is defined [here](https://github.com/geopython/pygeoapi/blob/master/pygeoapi/process/hello_world.py#L36)
* it specifies specific inputs/outputs definitions like `name` of type `text/plain` (should be `string`?)
* the generated `openapi.yaml` only refers to the generic OGC WPS REST definition `execute.yaml`

I think i.s.o. `execute.yaml` the pygeoapi-generated OpenAPI-specific inputs/outputs for `hello-world` need to be defined from the `PROCESS_METADATA` similar how `example` is used. (and those defs need to be validated against the OGC WPS REST spec).

**Comment by tomkralidis** [comment]
> Maybe the validator https://github.com/p1c2u/openapi-spec-validator is more strict?

Is there an authoritative validator?  For me, if it works in Swagger, it's fine (at this point).  Swagger editor/UI is indeed interpreting the `execute.yaml` as expected.

> it specifies specific inputs/outputs definitions like name of type text/plain (should be string?)

This is what we see currently in the 52 North implementation, and should also get clearer in a couple of weeks

> it specifies specific inputs/outputs definitions like name of type text/plain (should be string?)

This is in the spirit of WPS 2.0 REST.  This is sure to be distilled / clarified at the OGC API Hackathon

**Comment by justb4** [comment]
> > Maybe the validator https://github.com/p1c2u/openapi-spec-validator is more strict?
> 
> Is there an authoritative validator? For me, if it works in Swagger, it's fine (at this point). Swagger editor/UI is indeed interpreting the `execute.yaml` as expected.
> 
I expect OGC would be authoritative at some point (for OGC OpenAPI specs). IMHO https://github.com/p1c2u/openapi-spec-validator merely does a JSON-schema validation, similar to XML XSD validation e.g. with `lxml` which delegates even to different underlying parsers like `libxml2` depending on the installation. So multiple validators may do. I tried others like:
https://apidevtools.org/swagger-parser/online/ which gave similar errors on `execute.yaml` (unless I disabled following external `$refs`:
![image](https://user-images.githubusercontent.com/582630/59191532-12dc1a00-8b80-11e9-9034-e47070f6f5a6.png)
There' many others, few online: https://openapi.tools/

> > it specifies specific inputs/outputs definitions like name of type text/plain (should be string?)
> 
> This is what we see currently in the 52 North implementation, and should also get clearer in a couple of weeks
> 
Yes I am aware that the OGC WPS REST spec is derived from the 52North one.

> > it specifies specific inputs/outputs definitions like name of type text/plain (should be string?)
> 
> This is in the spirit of WPS 2.0 REST. This is sure to be distilled / clarified at the OGC API Hackathon

Still I think we should somehow specify the `requestBody` compliant with `execute.yaml`, not refer to it. There e.g. `outputs` is required.  And a user needs to know which `inputs` needs to be specified, not just an `example`. But again this whole domain is relatively new to me...

Also starts me thinking of a more modular approach to `pygeoapi`: a small kernel implementation and for each OGC API a module. Both in software and configuration. e.g. An implementation may only want to use `Processes` and not be bothered with `Features` (or the other way around).

**Comment by tomkralidis** [comment]
> Also starts me thinking of a more modular approach to pygeoapi: a small kernel implementation and for each OGC API a module. Both in software and configuration. e.g. An implementation may only want to use Processes and not be bothered with Features (or the other way around).

Agree.  I think the user configuration should be enough to set off what gets turned on (or not)?

**Comment by justb4** [comment]
> > Also starts me thinking of a more modular approach to pygeoapi: a small kernel implementation and for each OGC API a module. Both in software and configuration. e.g. An implementation may only want to use Processes and not be bothered with Features (or the other way around).
> 
> Agree. I think the user configuration should be enough to set off what gets turned on (or not)?


For now, yes, was thinking longer-term even in software distributions with only modules/formats needed, bit like MapServer and GDAL though in Python we can be more flexible (on-the-fly i.s.o. compiled-in, within Stetl and GHC we do something similar with Plugins), but we have currently much more basic issues to work on.

To get back to the requestBody schema and looking at `execute.yaml`: I think the `pygeoapi`-generated OpenAPI schema for the Hello-World Process should be, including the example, something like:

```
    post:
      description: Hello World process
      parameters: []
      requestBody:
        content:
          application/json:
            example:
              inputs:
              - id: name
                input:
                  dataType:
                    name: text/plain
                  value: hi there
            schema:
              type: object
              required:
                - outputs
              properties:
                inputs:
                  type: array
                  items:
                    type: object
                    properties:
                      id:
                        type: string
                        pattern: '^name$'
                      input:
                        type: object
                        properties:
                          value:
                            type: string
                          dataType:
                            type: object
                            required:
                              - name
                            properties:
                              name:
                                type: string
                                pattern: '^text/plain$'
                    minItems: 1
                    maxItems: 1
                outputs:
                  type: array
                  items:
                    type: object
                    properties:
.
.
etc
```
This validates ok.  This schema should then additionally comply with [execute.yaml](https://raw.githubusercontent.com/opengeospatial/wps-rest-binding/master/core/openapi/schemas/execute.yaml).

So apart from validation tools etc, I think this is what `pygeoapi` should generate. Or am I completely missing something? I know WPS is very generic (in inputs/outputs), but looks like the `pygeoapi` `Processes` are specific in parameterization. How would a user otherwise know which `inputs` and `outputs` to specify?

**Comment by tomkralidis** [comment]
@justb4 this should be fixed given #718.  Can you test/validate?
